Automattic/jetpack
# Jetpack Block PHP Fixture Tests: Add test for the server rendered Business Hours block

Related to: 251-gh-Automattic/view-design

With the recent work to add in [block test fixtures](https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/extensions/shared/test/block-fixtures.md) we noticed a gap in the testing for blocks that use server rendering. In addition to testing that the block's JS save method generates consistent markup, we also need to ensure that the serialized markup continues to result in correctly rendered server-side markup, when the block is rendered in PHP.

To explore one approach for testing fixtures for server-rendered blocks, this PR proposes adding in a test for the Business Hours block, since it has a fairly straightforward server rendering approach in that it doesn't call to any outside endpoints to generate its markup.

The approach I've explored involves:

The Business Hours block PHP test generates corresponding `*.server-rendered.html` fixtures for each of the block's `*.serialized.html` fixtures if they're not present. These generated fixtures are intended to be committed to the repo.

When matching generated fixtures exist, the PHP test checks the server-rendered `render_block` results against this fixture.

While working on writing or updating tests, a developer can run `yarn docker:phpunit --testsuite=extensions` to either test the existing fixtures, or if fixtures are missing, generate them.

#### Changes proposed in this Pull Request:

* For the Business Hours block, load all serialized fixtures matching the pattern `*.serialized.html`
* For each of these fixtures, check to see if a matching `*.server-rendered.html` fixture exists.
* If not, create one by passing the `*.serialized.html` fixture to `render_blocks` and save to disk. Then error out of the test to flag to the developer that the fixture file has been created and will need to be committed.
* If a fixture file does exist, then check to see that the fixture file's markup matches what's returned by calling `render_blocks` with the serialized markup for the block.

#### Jetpack product discussion



If this approach seems viable, then I can update the block fixtures documentation with a link to this example in a separate PR, and we can look at using this approach for other server-rendered blocks, too.

#### Does this pull request change what data or activity we track or use?



#### Testing instructions:






* Ensure the PHP tests are passing in CI

##### Testing locally

1. Check out this change locally, and following the instructions in this repo, set up your local Docker development environment and run the following from the root directory of this repo:

```
yarn docker:phpunit --testsuite=extensions
```

2. Delete the generated server-rendered fixtures included in this PR, and run the above command again. You should see that the test throws an error, and reports back the fixtures that have been generated:

![image](https://user-images.githubusercontent.com/14988353/117107718-633c7080-adc5-11eb-86fd-63ab6586706b.png)

3. Run the command again, and the tests should pass
4. Manually make a change to one of the server-rendered fixtures files and re-run the testing command. It should fail, complaining that the strings don't match. But, it should also tell you _which_ fixture it failed on with a (hopefully) friendly message than just the strings not matching:

![image](https://user-images.githubusercontent.com/14988353/117108469-8ddaf900-adc6-11eb-9e58-f9cda29d5232.png)

Also, a question for you! Does this approach to server rendered fixtures sound reasonable? Is there anything missing, or is there another approach that you think would work better instead?
